justfile

```diff
@@ -0,0 +1,138 @@
+
+target_dir := "target"
+doc_dir := "doc"
+doc_assets_dir := doc_dir + "/assets"
+screenshots_dir := target_dir + "/screenshots"
+
+#----------
+# Building
+#----------
+
+build: check-formatting test test-all build-benches build-simulator check-drawing-examples check-readmes generate-docs check-links
+
+# Build the benches
+build-benches:
+    cargo bench --no-run
+
+# Build the simulator
+build-simulator:
+    cd simulator; \
+    cargo build --release --no-default-features
+
+# Run cargo test in release mode
+test:
+    cargo test --release
+
+# Run cargo test in release mode with all features enabled
+test-all:
+    cargo test --release --all-features
+
+# Check the formatting
+check-formatting:
+    cargo fmt --all -- --check
+
+# Cross compiles embedded-graphics, tinytga and tinybmp for a target
+build-target target *args:
+    cargo build -p embedded-graphics --target {{target}} {{args}}
+    cd embedded-graphics; cargo build -p embedded-graphics --target {{target}} --all-features {{args}}
+
+    cargo build -p tinytga --target {{target}} {{args}}
+    cargo build -p tinytga --target {{target}} --all-features {{args}}
+
+    cargo build -p tinybmp --target {{target}} {{args}}
+    cargo build -p tinybmp --target {{target}} --all-features {{args}}
+
+#------
+# Docs
+#------
+
+# Generates the docs
+generate-docs:
+    cargo clean --doc
+    cargo doc --all-features
+
+# Runs linkchecker on the docs
+check-links:
+    linkchecker --check-extern --ignore-url=^http \
+        target/doc/embedded_graphics/index.html \
+        target/doc/tinybmp/index.html \
+        target/doc/tinytga/index.html \
+        target/doc/embedded_graphics_simulator/index.html
+
+#----------------------
+# README.md generation
+# ---------------------
+
+# Generate README.md for a single crate
+generate-readme crate: (_build-readme crate)
+    cp {{target_dir}}/{{crate}}_README.md {{crate}}/README.md
+
+# Check README.md for a single crate
+@check-readme crate: (_build-readme crate)
+    diff -q {{target_dir}}/{{crate}}_README.md ./{{crate}}/README.md || ( \
+        echo -e "\033[1;31mError:\033[0m README.md for {{crate}} needs to be regenerated."; \
+        echo -e "       Run 'just generate-readme {{crate}}' to regenerate.\n"; \
+        exit 1 \
+    )
+
+# Generate README.md for all crates
+generate-readmes:
+    for crate in {{crates}}; do just generate-readme $crate; done
+
+# Checks README.md for all crates
+check-readmes:
+    for crate in {{crates}}; do just check-readme $crate; done
+
+# Builds README.md for a single crate
+_build-readme crate:
+    #!/usr/bin/env bash
+    set -e -o pipefail
+    mkdir -p {{target_dir}}/readme
+    echo "Building README.md for {{crate}}"
+    cargo readme -r {{crate}} | sed -E -f filter_readme.sed > {{target_dir}}/{{crate}}_README.md
+
+#----------
+# Examples
+#----------
+
+# Generates the drawing examples screenshots and markdown file
+generate-drawing-examples:
+    #!/usr/bin/env bash
+    set -e -o pipefail
```

Comment:
I'm not sure why, but the `-o pipefail` here makes this command fail in CI with the following error

```
thread 'main' panicked at 'failed printing to stdout: Broken pipe (os error 32)', src/libstd/io/stdio.rs:792:9
```

Reply:
We could just remove it, but that might allow some false-positives to pass in CI.

Reply:
I just noticed this in the build log: "error: toolchain 'nightly-x86_64-unknown-linux-gnu' is not installed"

Reply:
Can you link the build that failed with that error?

Reply:
Oh doy I found it, never mind. Should've just scrolled up...